tree/ntuple/v7/inc/ROOT/RNTupleMetrics.hxx

```diff
@@ -0,0 +1,253 @@
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT7_RNTupleMetrics
+#define ROOT7_RNTupleMetrics
+
+#include <TError.h>
+
+#include <atomic>
+#include <chrono>
+#include <cstdint>
+#include <ctime> // for CPU time measurement with clock()
+#include <memory>
+#include <ostream>
+#include <string>
+#include <vector>
+#include <utility>
+
+namespace ROOT {
+namespace Experimental {
+namespace Detail {
+
+// clang-format off
+/**
+\class ROOT::Experimental::Detail::RNTuplePerfCounter
+\ingroup NTuple
+\brief A performance counter with a name and a unit, which can be activated on demand
+
+Derived classes decide on the counter type and implement printing of the value.
+*/
+// clang-format on
+class RNTuplePerfCounter {
+private:
+   std::string fName;
+   std::string fUnit;
+   std::string fDescription;
+   bool fIsEnabled = false;
+
+public:
+   RNTuplePerfCounter(const std::string &name, const std::string &unit, const std::string &desc)
+      : fName(name), fUnit(unit), fDescription(desc) {}
+   void Enable() { fIsEnabled = true; }
+   bool IsEnabled() const { return fIsEnabled; }
+   std::string GetName() const { return fName; }
+   std::string GetDescription() const { return fDescription; }
+   std::string GetUnit() const { return fUnit; }
+
+   virtual std::string ToString() const = 0;
+};
+
+
+// clang-format off
+/**
+\class ROOT::Experimental::Detail::RNTuplePlainCounter
+\ingroup NTuple
+\brief A non thread-safe integral performance counter
+*/
+// clang-format on
+class RNTuplePlainCounter : public RNTuplePerfCounter {
+private:
+   std::int64_t fCounter = 0;
+
+public:
+   RNTuplePlainCounter(const std::string &name, const std::string &unit, const std::string &desc)
+      : RNTuplePerfCounter(name, unit, desc)
+   {
+   }
+
+   void Inc() { ++fCounter; }
+   void Dec() { --fCounter; }
+   void Add(int64_t delta) { fCounter += delta; }
+   int64_t GetValue() const { return fCounter; }
+   void SetValue(int64_t val) { fCounter = val; }
+   std::string ToString() const override { return std::to_string(fCounter); }
+};
+
+
+// clang-format off
+/**
+\class ROOT::Experimental::Detail::RNTupleAtomicCounter
+\ingroup NTuple
+\brief A thread-safe integral performance counter
+*/
+// clang-format on
+class RNTupleAtomicCounter : public RNTuplePerfCounter {
+private:
+   std::atomic<std::int64_t> fCounter = 0;
+
+public:
+   RNTupleAtomicCounter(const std::string &name, const std::string &unit, const std::string &desc)
+      : RNTuplePerfCounter(name, unit, desc) { }
+
+   void Inc() {
+      if (IsEnabled())
+         ++fCounter;
+   }
+   void Dec() {
+      if (IsEnabled())
+         --fCounter;
+   }
+   void Add(int64_t delta) {
+      if (IsEnabled())
```

Comment:
For performance reason you might want to (eventually) add R__ALWAYS_INLINE and R__unlikely:
``` 
R__ALWAYS_INLINE
void Add(int64_t delta) {
      if (R__unlikely(IsEnabled()))
         fCounter += delta;
}